util/sst_file_manager_impl.cc

```diff
@@ -118,15 +150,44 @@ bool SstFileManagerImpl::EnoughRoomForCompaction(
     }
   }
 
-  if (max_allowed_space_ != 0 &&
-      (size_added_by_compaction + cur_compactions_reserved_size_ +
-           total_files_size_ + compaction_buffer_size_ >
-       max_allowed_space_)) {
-    return false;
-  }
   // Update cur_compactions_reserved_size_ so concurrent compaction
   // don't max out space
+  // needed_headroom_ is based on current size reserved by compactions,
+  // minus any files created by running compactions as they would count
+  // against the reserved size.
   cur_compactions_reserved_size_ += size_added_by_compaction;
+  size_t needed_headroom =
+      cur_compactions_reserved_size_ - in_progress_files_size_ +
+      (compaction_buffer_size_ > 0 ? compaction_buffer_size_
+                                   : reserved_disk_buffer_);
+  if (max_allowed_space_ != 0 &&
+      (needed_headroom + total_files_size_ > max_allowed_space_)) {
+    cur_compactions_reserved_size_ -= size_added_by_compaction;
+    return false;
+  }
+
+  // Implement more aggressive checks only if this DB instance has already
+  // seen a NoSpace() error. This is tin order to contain a single potentially
+  // misbehaving DB instance and prevent it from slowing down compactions of
+  // other DB instances
+  if (CheckFreeSpace() && bg_error == Status::NoSpace()) {
+    auto fn =
+        TableFileName(cfd->ioptions()->cf_paths, inputs[0][0]->fd.GetNumber(),
+                      inputs[0][0]->fd.GetPathId());
+    uint64_t free_space = 0;
+    env_->GetFreeSpace(fn, &free_space);
+    if (free_space < needed_headroom) {
+      // We hit the condition of not enough disk space
+      cur_compactions_reserved_size_ -= size_added_by_compaction;
+      ROCKS_LOG_ERROR(logger_, "free space [%d bytes] is less than "
+          "needed headroom [%d bytes]\n", free_space, needed_headroom);
+      return false;
+    }
+  } else {
+    // Take a snapshot of cur_compactions_reserved_size_ for when we encounter
+    // a NoSpace error.
+    free_space_trigger_ = cur_compactions_reserved_size_;
```

Comment:
I have a worry using this one.
For example, if a user runs universal compaction, and has 200 shards, and it is out-of-memory. Eventually, full compaction of all the shards will reserve the size equivalent to the whole shard size, and fail. The compaction trigger would be equal to the total size of the 200 shards. Then users need to drop the disk space to lower than 50% to make it recover. This is going to be much more than what is needed. Is it a valid worry?

Reply:
The ```cur_compactions_reserved_size_``` is the space reserved by currently running compactions. In your example, would the user tune ```max_background_jobs``` to limit the number of concurrent compactions? If they set it to a high number, they will always run the risk of doubling the space used. This check may, in fact benefit that case, as it does not stop background work but rather just slows it down. If ```EnoughRoomForCompaction``` returns false, the compaction is retried after some time. So it would end up throttling concurrent compactions to some lower number that can actually be accommodated by the available disk space.